src/main/cpp/src/hyper_log_log_plus_plus.cu

```diff
@@ -0,0 +1,952 @@
+ * It uses 2 phase merges: partial merge and final merge
+ *
+ * This function splits input into multiple segments with each segment has
+ * num_hashs_per_thread items. The input is sorted by group labels, each segment
+ * contains one or more consecutive groups. Each thread handles one segment with
+ * num_hashs_per_thread items in it:
+ * - Scan all the items in the segment, update the max value.
+ * - Output max value into registers_output_cache for the previous group when
+ * meets a new group.
+ * - Output max value into registers_thread_cache when reach the last item in
+ * the segment.
+ *
+ * In this way, we can save memory usage, cache less intermidate sketches
+ * (num_hashs / num_hashs_per_thread) sketches.
+ * num_threads = div_round_up(num_hashs, num_hashs_per_thread).
+ *
+ * Note: Must exclude null hash values from computing HLLPP sketches.
+ *
+ * e.g.: num_registers_per_sketch = 512 and num_hashs_per_thread = 4;
+ *
+ * Input is hashs, compute and get pair: register index -> register value
+ *
+ *   reg_index, reg_value, group_label
+ * [
+ * ---------- segment 0 begin --------------------------
+ *    (0,            1),          g0
+ *    (0,            2),          g0
+ * // meets new group g1, save result for group g0 into registers_output_cache
+ *    (1,            1),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ *    (1,            9),          g1
+ * ---------- segment 1 begin --------------------------
+ *    (1,            1),          g1
+ *    (1,            1),          g1
+ *    (1,            5),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ *    (1,            1),          g1
+ * ---------- segment 2 begin --------------------------
+ *    (1,            1),          g1
+ *    (1,            1),          g1
+ *    (1,            8),          g1
+ * // outputs result at segemnt end for this thread to registers_thread_cache
+ * // assumes meets new group when at the end, save to registers_output_cache
+ *    (1,            1),          g1
+ * ]
+ * Output e.g.:
+ *
+ * group_labels_thread_cache:
+ * [
+ *   g1
+ *   g1
+ *   g1
+ * ]
+ * Has num_threads rows.
+ *
+ * registers_thread_cache:
+ * [
+ *    512 values: [0, 9, 0, ... ] // register values for group 1
+ *    512 values: [0, 5, 0, ... ] // register values for group 1
+ *    512 values: [0, 8, 0, ... ] // register values for group 1
+ * ]
+ * Has num_threads rows, each row is corresponding to
+ * `group_labels_thread_cache`
+ *
+ * registers_output_cache:
+ * [
+ *    512 values: [2, 0, 0, ... ] // register values for group 0
+ *    512 values: [0, 8, 0, ... ] // register values for group 1
+ * ]
+ * Has num_groups rows.
+ *
+ * The next kernel will merge the registers_output_cache and
+ * registers_thread_cache and get the final result.
+ */
+template <int num_hashs_per_thread>
+CUDF_KERNEL void partial_group_sketches_from_hashs_kernel(
```

Comment:
Now it is recommended to add `__launch_bounds__` to all kernels to avoid any potential issues due to compiler regression that compile the kernels to use much more registers.